Allow emailPrefix to include '%(repo_shortname)s'

Irfy opened the pull request:
The default emailPrefix is `'[%s]' % (self.get_repo_name(),)` but there seems to be no way of referring to the repo name in a custom emailPrefix.

This patch allows a custom emailPrefix to substitute `%(repo_shortname)s` with the repo name.

moy commented:
Thanks for the PR. Did you read the discussion on  #19, about a similar request?

I like the idea of your PR, but I find it a bit too specific. A more general solution would be to allow overriding all template strings via the Git configuration file. Then one would override `*_SUBJECT_TEMPLATE` templates and set the prefix there. But that would force the user to set 4 template strings, while your solution is much simpler. So, it seems to be the way to go, even if we later allow custom template strings from the config.

Still, why restrict the variable expansion to `%(repo_shortname)s`? Why not allow all variables available in template strings?

Also, this lacks documentation and test.

moy commented on git-multimail/git_multimail.py:
```diff
@@ -2677,11 +2677,18 @@ def get_emailprefix(self):
+        try:
+            return emailprefix % {'repo_shortname': self.get_repo_shortname()}
+        except:
+            sys.stderr.write(
```
There's a new logging mechanism, please use stg like

            self.get_logger().error(...)

instead.

moy commented on git-multimail/git_multimail.py:
```diff
@@ -2677,11 +2677,18 @@ def get_emailprefix(self):
+                '*** %s' % (sys.exc_info()[1]) +
+                "*** Only the '%(repo_shortname)s' placeholder is allowed\n"
+                )
+            sys.exit(1)
```
We normally avoid `sys.exit`, to let the caller catch potential exception. This looks like a `ConfigurationException`.

Irfy commented:
- replaced stderr.write with logging
- added a test for the successful case. I'd need some pointers for the failing case -- I'd like to write one for that too (I had already caught myself with a typo in that code path)
- added documentation (is this the right way?)

About allowing replacing all available variables: I went for the most straightforward thing that will cover the vast majority of uses, which is to reference the repo_shortname. If we wanted to allow any of the COMPUTED_KEYS, we'd have to either:
- duplicate the get_values() implementation in some way, or
- modify it so that it calls itself from within get_emailprefix without recursing further, or
- specially handle the emailprefix by allowing substitution for it from the generated values in get_values, at the end of get_values
- allow all values one chance to substitute any placeholders that were just generated. But then do we detect crossreferences like emailPrefix=%(repo_shortname)s and repo_shortname=%(emailPrefix)?

I think anything more complex than this is unnecessary, and should be done when someone needs it, i.e. opens an issue with a legitimate/defensible problem description, or submits a PR like me, justifying their position.

In my case, I just needed to have git commit email subjects have the same prefix as our team's svn commit email subjects, nothing more. The other parameters are either already in the templates, or make no sense in a subject line, IMHO.

moy commented:
> I think anything more complex than this is unnecessary, and should be done when someone needs it

OK. The important part to me is that the feature can be extended later without backward compatibility issue. The code looked a bit too ad-hoc to me, but at least it is simple, which I like.

moy commented on git-multimail/README:
```diff
@@ -447,7 +447,9 @@ multimailhook.emailPrefix
     the repository in square brackets; e.g., ``[myrepo]``.  Set this
-    value to the empty string to suppress the email prefix.
+    value to the empty string to suppress the email prefix. You may
+    use the placeholder %(repo_shortname)s for the short name of the
```
Don't forget to escape literal code with double-backticks. I can do it while applying.

moy commented:
I fixed the minor formatting issue in README and added a negative test (7d6d34afe09299f006fbe9c541cf6a896b71b1ee). Thanks for the contribution and the quick response to review!

Irfy commented:
Great!
